Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.m

```diff
@@ -0,0 +1,311 @@
+ * MXMetricManager subscribers. Also initializes the promise that we'll use to ensure that any
+ * MetricKit report files are included in Crashylytics fatal reports. If no crash occurred on the
+ * last run of the app, this promise is immediately resolved so that the upload of any nonfatal
+ * events can proceed.
+ */
+- (void)registerMetricKitManager {
+  [[MXMetricManager sharedManager] addSubscriber:self];
+  _metricKitDataAvailable = [FBLPromise pendingPromise];
+
+  // If there was no crash on the last run of the app, then we aren't expecting a MetricKit
+  // diagnostic report and should resolve the promise immediately.
+  NSString *crashedMarkerFileName = [NSString stringWithUTF8String:FIRCLSCrashedMarkerFileName];
+  NSString *crashedMarkerFileFullPath =
+      [[_fileManager rootPath] stringByAppendingPathComponent:crashedMarkerFileName];
+  if (![_fileManager fileExistsAtPath:crashedMarkerFileFullPath]) {
+    [_mutex lock];
+    [_metricKitDataAvailable fulfill:nil];
+    [_mutex unlock];
+  }
+}
+
+/*
+ * This method receives diagnostic payloads from MetricKit whenever a fatal or nonfatal MetricKit
+ * event occurs. If a fatal event, this method will be called when the app restarts. Since we're
+ * including a MetricKit report file in the Crashlytics report to be sent to the backend, we need
+ * to make sure that we process the payloads and write the included information to file before
+ * the report is sent up. If this method is called due to a nonfatal event, it will be called
+ * immediately after the event. Since we send nonfatal events on the next run of the app, we can
+ * write out the information but won't need to resolve the promise.
+ */
+- (void)didReceiveDiagnosticPayloads:(NSArray<MXDiagnosticPayload *> *)payloads {
+  NSUInteger payloadCount = [payloads count];
+  FIRCLSDebugLog(@"[Crashlytics:Crash] received %lu diagnostic payloads from MetricKit",
+                 payloadCount);
+
+  if ([payloads count] == 0) {
+    return;
+  }
+
+  MXDiagnosticPayload *diagnosticPayload = [payloads objectAtIndex:0];
+  if (!diagnosticPayload) {
+    return;
+  }
+
+  // TODO: Time stamp information is only available in begin and end time periods. Hopefully this is
+  // updated with iOS 15.
+  NSTimeInterval beginSecondsSince1970 = [diagnosticPayload.timeStampBegin timeIntervalSince1970];
+  NSTimeInterval endSecondsSince1970 = [diagnosticPayload.timeStampEnd timeIntervalSince1970];
+
+  // Get file path for the active reports directory.
+  NSString *activePath = [[_fileManager activePath] stringByAppendingString:@"/"];
+
+  // If newestReportID is nil or the file no longer exists, then this callback function was
+  // triggered for nonfatal MetricKit events.
+  NSString *newestUnsentReportID =
+      [_existingReportManager.newestUnsentReport.reportID stringByAppendingString:@"/"];
+  NSString *metricKitReportFile;
+  BOOL fatal =
+      (newestUnsentReportID != nil) && ([_fileManager fileExistsAtPath:metricKitReportFile]);
```

Comment:
I don't think `newestUnsentReportID` will ever be nil because you append a slash to it. Can just check `self.existingReportManager.newestUnsentReport` is nil

Reply:
So it turns out that it can be nil, the append fails if the original string is nil